### PR TITLE
CHT test: Reorg with 550 blocks and presentation of another chain - issue 27

### PR DESCRIPTION
**Pull request by fenix2222**
Checkpoints are disabled. Chain tip is at header 5. Reorganisation is initiaised with max reorg of 500 blocks plus extra 50 blocks. We start syncing until block 490. Peer 2 presents the alternative chain with 2 headers after fork point at header 5. We then you join the rest of 60 blocks. ConsensusTipChanged should return identifier of the second peer at block number 505.

See #1321 (comment)

Preferred reviewers: @Aprogiena or @noescape00

### CODE REVIEW

File: src/Stratis.Bitcoin.Tests/Consensus/ConsensusTestExtensions.cs [hunk 1]
```diff
@@ -26,5 +26,21 @@ public static ChainedHeader GetPeerTipChainedHeaderByPeerId(this ChainedHeaderTr
         {
             return chainedHeaderTree.GetChainedHeadersByHash()[chainedHeaderTree.GetPeerTipsByPeerId()[peer]];
         }
+
+        public static List<ChainedHeader> ToList(this ChainedHeader chainedHeader, int headersToTake)
+        {
+            var list = new List<ChainedHeader>();
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ConsensusTestExtensions.cs]: You can initialise with capacity.

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 2]
```diff
@@ -1063,5 +1063,65 @@ public void BlockDataDownloadedIsCalled_ForValidBlocksAfterFv_ResultShouldBeTrue
             resultForH5.Should().BeTrue();
             chainTipH5.Block.Should().NotBeNull();
         }
+
+        /// <summary>
+        /// Issue 26 @  Checkpoints are disabled. Chain tip is at header 5. Reorganisation is initiaised with max reorg of 500 blocks plus extra 50 blocks.
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: It is supposed to be 27

Comment (Aprogiena) [reply src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: And there actually is no reorganisation in this test, it is more about discarding a chain that can no longer be viable in the future.

Comment (Aprogiena) [reply src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Or more correctly, it is about identifying the peers that presents such chains.

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 3]
```diff
@@ -1063,5 +1063,65 @@ public void BlockDataDownloadedIsCalled_ForValidBlocksAfterFv_ResultShouldBeTrue
...
+            List<ChainedHeader> listOfChainAChainHeaders = chainATip.ToList(maxReorg + 50);
+            cht.ConnectNewHeaders(1, listOfChainABlockHeaders);
+            
+            // Sync 490 blocks from chain A.
+            for (int i = 0; i < maxReorg - 10; i++)
+            {
+                ChainedHeader currentChainTip = listOfChainAChainHeaders[i];
+                cht.BlockDataDownloaded(currentChainTip, currentChainTip.Block);
+                cht.PartialValidationSucceeded(currentChainTip, out bool reorgRequired);
+                List<int> tips = cht.ConsensusTipChanged(currentChainTip);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: You are not using this list anywhere, but we could actually check that it is null as it should be.

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 4]
```diff
@@ -1063,5 +1063,65 @@ public void BlockDataDownloadedIsCalled_ForValidBlocksAfterFv_ResultShouldBeTrue
...
+            {
+                ChainedHeader currentChainTip = listOfChainAChainHeaders[i];
+                cht.BlockDataDownloaded(currentChainTip, ctx.CreateBlock());
+                cht.PartialValidationSucceeded(currentChainTip, out bool reorgRequired);
+                List<int> tips = cht.ConsensusTipChanged(currentChainTip);
+                if (currentChainTip.Height == maxReorg + initialChainSize)
+                {
+                    tips.Should().HaveCount(1);
+                    tips[0].Should().Be(2);
+                }
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: He also we could have an else branch testing the emptiness. Also the list does not contain any tips or headers, so it's name is misleading. It is list of peers who present such chains (see above).

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 5]
```diff
@@ -1063,5 +1063,65 @@ public void BlockDataDownloadedIsCalled_ForValidBlocksAfterFv_ResultShouldBeTrue
...
+            // Example: h1=h2=h3=h4=(h5)=b7=b8.
+            const int chainBExtension = 2;
+            ChainedHeader chainBTip = ctx.ExtendAChain(chainBExtension, initialChainTip);
+
+            // Chain B is presented by peer 2.
+            List<BlockHeader> listOfChainBHeaders = ctx.ChainedHeaderToList(chainBTip, chainBExtension);
+            cht.ConnectNewHeaders(2, listOfChainBHeaders);
+
+            // Continue syncing remaining blocks from chain A.
+            // TODO: add asserts
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: We can probably remove this now.

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 6]
```diff
@@ -1063,5 +1063,70 @@ public void BlockDataDownloadedIsCalled_ForValidBlocksAfterFv_ResultShouldBeTrue
...
+                if (currentChainTip.Height == maxReorg + initialChainSize)
+                {
+                    peerIds.Should().HaveCount(1);
+                    peerIds[0].Should().Be(2);
+                }
+                else
+                {
+                    peerIds.Should().BeEmpty();
+                }
+
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Empty line

File: src/Stratis.Bitcoin.Tests/Consensus/ConsensusTestExtensions.cs [hunk 7]
```diff
@@ -1,4 +1,5 @@
 ﻿using System.Collections.Generic;
+using System.Linq;
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ConsensusTestExtensions.cs]: Still needed?

File: src/Stratis.Bitcoin.Tests/Consensus/ConsensusTestExtensions.cs [hunk 8]
```diff
@@ -26,5 +27,19 @@ public static ChainedHeader GetPeerTipChainedHeaderByPeerId(this ChainedHeaderTr
         {
             return chainedHeaderTree.GetChainedHeadersByHash()[chainedHeaderTree.GetPeerTipsByPeerId()[peer]];
         }
+
+        public static ChainedHeader[] ToArray(this ChainedHeader chainedHeader, int headersToTake)
+        {
+            var headers = new ChainedHeader[headersToTake];
+            ChainedHeader current = chainedHeader;
+
+            for (int i = headersToTake - 1; i >= 0 && current != null; i--)
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ConsensusTestExtensions.cs]: Brackets please, see code style

**Comment by Aprogiena** [comment]
Looks good to me now

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 9]
```diff
@@ -1063,5 +1063,72 @@ public void BlockDataDownloadedIsCalled_ForValidBlocksAfterFv_ResultShouldBeTrue
...
+            // Create new chain B with 2 headers after the fork point.
+            // Example: h1=h2=h3=h4=(h5)=b7=b8.
+            const int chainBExtension = 2;
+            ChainedHeader chainBTip = ctx.ExtendAChain(chainBExtension, initialChainTip);
+
+            // Chain B is presented by peer 2.
+            List<BlockHeader> listOfChainBHeaders = ctx.ChainedHeaderToList(chainBTip, chainBExtension);
+            Action connectAction = () => cht.ConnectNewHeaders(2, listOfChainBHeaders);
+            connectAction.Should().Throw<MaxReorgViolationException>();
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: If the finalised block height is correctly set up as mentioned above, this should not throw

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 10]
```diff
@@ -1063,5 +1063,72 @@ public void BlockDataDownloadedIsCalled_ForValidBlocksAfterFv_ResultShouldBeTrue
...
+
+            // Continue syncing remaining blocks from chain A.
+            for (int i = maxReorg - 10; i < maxReorg + 50; i++)
+            {
+                ChainedHeader currentChainTip = chainAChainHeaders[i];
+                cht.BlockDataDownloaded(currentChainTip, ctx.CreateBlock());
+                cht.PartialValidationSucceeded(currentChainTip, out bool reorgRequired);
+                ctx.FinalizedBlockMock.Setup(m => m.GetFinalizedBlockHeight()).Returns(currentChainTip.Height);
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: Same as above

File: src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs [hunk 11]
```diff
@@ -1063,5 +1063,72 @@ public void BlockDataDownloadedIsCalled_ForValidBlocksAfterFv_ResultShouldBeTrue
...
+                List<int> peerIds = cht.ConsensusTipChanged(currentChainTip);
+                if (currentChainTip.Height == maxReorg + initialChainSize)
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs]: I have just realised that we do expect nonempty list to be returned not only at this block, but also for all blocks above it

**Comment by noescape00** [comment]
@Aprogiena merge when you're ready.

**Pull request merged**